src/hotspot/share/opto/macroArrayCopy.cpp

```diff
@@ -204,53 +204,46 @@ void PhaseMacroExpand::generate_limit_guard(Node** ctrl, Node* offset, Node* sub
 void PhaseMacroExpand::generate_partial_inlining_block(Node** ctrl, MergeMemNode** mem, const TypePtr* adr_type,
                                                        RegionNode** exit_block, Node** result_memory, Node* length,
                                                        Node* src_start, Node* dst_start, BasicType type) {
-  const TypePtr *src_adr_type = _igvn.type(src_start)->isa_ptr();
-  Node* inline_block = nullptr;
-  Node* stub_block = nullptr;
-
-  int const_len = -1;
-  const TypeInt* lty = nullptr;
-  uint shift  = exact_log2(type2aelembytes(type));
-  if (length->Opcode() == Op_ConvI2L) {
-    lty = _igvn.type(length->in(1))->isa_int();
-  } else  {
-    lty = _igvn.type(length)->isa_int();
-  }
-  if (lty && lty->is_con()) {
-    const_len = lty->get_con() << shift;
+  int inline_limit = ArrayOperationPartialInlineSize / type2aelembytes(type);
+
+  const TypeLong* length_type = _igvn.type(length)->isa_long();
```

Comment:
Any particular benefit in eagerly pruning the block? It duplicates post-expansion GVN checks of the branch condition. (If it were normal parsing with prompt GVN analysis, you could detect the branch is dead right after `generate_guard` call.) 

Alternatively, the checks are equivalent to checking that join of `length_type` with `[0...inline_limit]` is not empty. But I prefer to let GVN handle it.

Reply:
I think it is a trivial check and it is much more efficient than creating a bunch of nodes and removing them later.

Reply:
Ok, then please, turn it into:
```
const TypeLong* inline_range = TypeLong::make(0, inline_limit);
if (length_type->join(inline_range) == Type::TOP) {
  return;
}
```
Or even encapsulate the check in `ArrayCopyNode::get_partial_inline_vector_lane_count()` and guard the subsequent checks with `lane_count > 0`.

Otherwise, the patch looks good.

Reply:
Thanks a lot, done!